Source/WebCore/css/CSSBasicShapes.cpp

```diff
@@ -191,6 +191,116 @@ bool CSSEllipseValue::equals(const CSSEllipseValue& other) const
+{
+    return adoptRef(*new CSSXywhValue(WTFMove(insetX), WTFMove(insetY), WTFMove(width), WTFMove(height), WTFMove(topLeftRadius), WTFMove(topRightRadius), WTFMove(bottomRightRadius), WTFMove(bottomLeftRadius)));
+}
+
+bool CSSXywhValue::equals(const CSSXywhValue& other) const
+{
+    return compareCSSValue(m_insetX, other.m_insetX)
+        && compareCSSValue(m_insetY, other.m_insetY)
+        && compareCSSValue(m_width, other.m_width)
+        && compareCSSValue(m_height, other.m_height)
+        && compareCSSValuePtr(m_topLeftRadius, other.m_topLeftRadius)
+        && compareCSSValuePtr(m_topRightRadius, other.m_topRightRadius)
+        && compareCSSValuePtr(m_bottomRightRadius, other.m_bottomRightRadius)
+        && compareCSSValuePtr(m_bottomLeftRadius, other.m_bottomLeftRadius);
+}
+
+static inline void updateCornerRadiusWidthAndHeight(const CSSValue* corner, String& width, String& height)
+{
+    if (!corner)
+        return;
+    width = corner->first().cssText();
+    height = corner->second().cssText();
+}
+
+static bool buildRadii(Vector<String>& radii, const String& topLeftRadius, const String& topRightRadius, const String& bottomRightRadius, const String& bottomLeftRadius)
+{
+    bool showBottomLeft = topRightRadius != bottomLeftRadius;
+    bool showBottomRight = showBottomLeft || (bottomRightRadius != topLeftRadius);
+    bool showTopRight = showBottomRight || (topRightRadius != topLeftRadius);
+
+    radii.append(topLeftRadius);
+    if (showTopRight)
+        radii.append(topRightRadius);
+    if (showBottomRight)
+        radii.append(bottomRightRadius);
+    if (showBottomLeft)
+        radii.append(bottomLeftRadius);
+
+    return radii.size() == 1 && radii[0] == "0px"_s;
+}
+
+static String buildXywhString(const String& insetX, const String& insetY, const String& width, const String& height,
+    const String& topLeftRadiusWidth, const String& topLeftRadiusHeight,
+    const String& topRightRadiusWidth, const String& topRightRadiusHeight,
+    const String& bottomRightRadiusWidth, const String& bottomRightRadiusHeight,
+    const String& bottomLeftRadiusWidth, const String& bottomLeftRadiusHeight)
+{
+    StringBuilder result;
+    result.append("xywh("_s, insetX, ' ', insetY, ' ', width, ' ', height);
+
+    if (!topLeftRadiusWidth.isNull() && !topLeftRadiusHeight.isNull()) {
+        Vector<String> horizontalRadii;
+        bool areDefaultCornerRadii = buildRadii(horizontalRadii, topLeftRadiusWidth, topRightRadiusWidth, bottomRightRadiusWidth, bottomLeftRadiusWidth);
+
+        Vector<String> verticalRadii;
+        areDefaultCornerRadii &= buildRadii(verticalRadii, topLeftRadiusHeight, topRightRadiusHeight, bottomRightRadiusHeight, bottomLeftRadiusHeight);
+
+        if (!areDefaultCornerRadii) {
+            result.append(" round"_s);
+
+            for (auto& radius : horizontalRadii)
+                result.append(' ', radius);
+
+            if (verticalRadii.size() != horizontalRadii.size()
```

Comment:
Why can't this just be `if (verticalRadii != horizontalRadii)`? `VectorComparer` is supposed to be an implementation detail of `Vector` and I am surprised you need to use it directly.